pkg/controller/machine_safety.go

```diff
@@ -68,6 +68,107 @@ func (c *controller) reconcileClusterMachineSafetyOvershooting(key string) error
 	return nil
 }
 
+// reconcileClusterMachineSafetyAPIServer checks control and target clusters
+// and checks if their APIServer's are reachable
+// If they are not reachable, they set a machineControllerFreeze flag
+func (c *controller) reconcileClusterMachineSafetyAPIServer(key string) error {
+	statusCheckTimeout := c.safetyOptions.MachineSafetyAPIServerStatusCheckTimeout.Duration
+	statusCheckPeriod := c.safetyOptions.MachineSafetyAPIServerStatusCheckPeriod.Duration
+
+	glog.V(3).Infof("reconcileClusterMachineSafetyAPIServer: Start")
+	defer glog.V(3).Infof("reconcileClusterMachineSafetyAPIServer: Stop")
+
+	if c.safetyOptions.MachineControllerFrozen {
+		// MachineController is frozen
+		if c.isAPIServerUp() {
+			// APIServer is up now, hence we need reset all machine health checks (to avoid unwanted freezes) and unfreeze
+			machines, err := c.machineLister.List(labels.Everything())
+			if err != nil {
+				glog.Warning(err)
+				return err
+			}
+			for _, machine := range machines {
+				if machine.Status.CurrentStatus.Phase == v1alpha1.MachineUnknown {
+					machine, err := c.controlMachineClient.Machines(c.namespace).Get(machine.Name, metav1.GetOptions{})
+					if err != nil {
+						glog.Warning(err)
+						return err
+					}
+
+					machine.Status.CurrentStatus = v1alpha1.CurrentStatus{
+						Phase:          v1alpha1.MachineRunning,
```

Comment:
what if the machine was in some other state when the API server went down. Should we still set the phase as "running"? Eg. What about "terminating" or "pending"?

Reply:
Ignore this comment, as we set the "unknown" state only if machine was "running".